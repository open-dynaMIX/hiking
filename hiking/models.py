import datetime
import os
from typing import List

import gpxpy
from sqlalchemy import (
    Column,
    Date,
    Float,
    Integer,
    Interval,
    String,
    Text,
    create_engine,
)
from sqlalchemy.orm import declarative_base, sessionmaker

from hiking.utils import DB_PATH, format_value, pretty_timedelta

Base = declarative_base()
engine = create_engine(
    f"sqlite:///{str(DB_PATH.absolute())}"
    if not os.environ.get("HIKING_TEST")
    else "sqlite://"  # , echo=True
)
Session = sessionmaker(bind=engine)
session = Session()


def create_tables():
    Base.metadata.create_all(engine)


class Hike(Base):
    __tablename__ = "hikes"

    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)
    body = Column(Text, nullable=True)
    date = Column(Date, nullable=False)
    distance = Column(Float, nullable=False)
    elevation_gain = Column(Integer, nullable=False)
    elevation_loss = Column(Integer, nullable=False)
    duration = Column(Interval, nullable=False)
    gpx_xml = Column(Text)

    FIELD_PROPS = {
        "id": {
            "pretty_name": "ID",
            "supported_calculations": [],
            "data_view": True,
            "calculated_value": False,
        },
        "date": {
            "pretty_name": "Date",
            "supported_calculations": [],
            "data_view": True,
            "calculated_value": False,
        },
        "name": {
            "pretty_name": "Name",
            "supported_calculations": [],
            "data_view": True,
            "calculated_value": False,
        },
        "body": {
            "pretty_name": "Body",
            "supported_calculations": [],
            "data_view": False,
            "calculated_value": False,
        },
        "distance": {
            "pretty_name": "➡ km",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "elevation_gain": {
            "pretty_name": "⬈ m",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "elevation_loss": {
            "pretty_name": "⬊ m",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "duration": {
            "pretty_name": "⏱ ",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "speed": {
            "pretty_name": "km/h",
            "supported_calculations": ["avg", "min", "max"],
            "data_view": True,
            "calculated_value": True,
        },
        "gpx": {
            "pretty_name": "GPX",
            "supported_calculations": [],
            "data_view": False,
            "calculated_value": False,
        },
    }

    _gpx = None

    @property
    def speed(self):
        return self.distance * 60 / (self.duration.seconds / 60)

    @property
    def gpx(self):
        if self.gpx_xml and not self._gpx:
            self._gpx = gpxpy.parse(self.gpx_xml)
        return self._gpx

    def get_pretty_value(
        self,
        attr: str,
    ):
        value = getattr(self, attr)
        return format_value(value, attr)

    def get_stats(self) -> List[str]:
        serialized = []
        for attr, config in self.FIELD_PROPS.items():
            if not config["data_view"]:
                continue

            value = self.get_pretty_value(attr)
            serialized.append(value)

        return serialized

    def get_detail_stats(self) -> dict:
        serialized = {}
        for f, config in self.FIELD_PROPS.items():
            if not config["data_view"]:
                continue
            value = getattr(self, f)
            if isinstance(value, datetime.timedelta):
                value = pretty_timedelta(value)
            elif f == "speed":
                value = round(value, 2)
            serialized[config["pretty_name"]] = value

        return serialized

    def load_gpx(self, gpx: str):
        gpx_obj = gpxpy.parse(gpx)

        data = {
            "date": gpx_obj.time.date() if gpx_obj.time else None,
            "name": gpx_obj.name,
            "distance": round(gpx_obj.length_3d() / 1000, 2),
            "elevation_gain": round(gpx_obj.get_uphill_downhill().uphill),
            "elevation_loss": round(gpx_obj.get_uphill_downhill().downhill),
            "duration": gpx_obj.get_duration(),
            "gpx_xml": gpx,
        }
        for attr, value in data.items():
            if value:
                setattr(self, attr, value)

    def save(self):
        if self.id is None:
            session.add(self)
        return session.commit()

    def delete(self):
        session.delete(self)
        return session.commit()

    def __str__(self) -> str:
        return f"<Hike - {self.name} - {self.date} - {self.distance}km>"

    def __repr__(self) -> str:
        return self.__str__()
