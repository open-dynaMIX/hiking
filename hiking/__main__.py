#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import logging
import sys

from hiking import commands
from hiking.arg_parsing import parse_arguments
from hiking.exceptions import HikingException, HikingJsonLoaderException
from hiking.import_export import JSON_IMPORT_EXAMPLE
from hiking.models import create_tables
from hiking.utils import DATA_HOME

logger = logging.getLogger(__name__)


def main():
    args = parse_arguments(sys.argv[1:])

    DATA_HOME.mkdir(exist_ok=True)
    create_tables()

    try:
        match args.command:
            case "show":
                commands.command_show(
                    args.ids,
                    args.daterange,
                    args.table_style,
                    args.order_key,
                    args.plot,
                )
            case "create" | "edit":
                commands.command_create_edit(pk=getattr(args, "id", None), gpx=args.gpx)
            case "delete":
                commands.command_delete(args.ids, args.all, args.force, args.quiet)
            case "import":
                commands.command_import(args.json_data)
            case "export":
                commands.command_export(
                    args.export_dir, args.ids, args.daterange, args.include_ids
                )

    except HikingJsonLoaderException as e:
        logger.warning(
            f"Invalid data in hiking.json: {e.args[0]}\n\nExpected format:\n{JSON_IMPORT_EXAMPLE}"
        )
    except HikingException as e:
        logger.error(f"Error: {e}")
    except (KeyboardInterrupt, EOFError):  # pragma: no cover
        sys.exit(0)


if __name__ == "__main__":  # pragma: no cover
    main()
