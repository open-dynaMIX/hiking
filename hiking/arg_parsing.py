import argparse
import datetime
import json
import os
from itertools import zip_longest
from pathlib import Path
from typing import List, Optional, Tuple

import gpxpy
from rich import box

from hiking.import_export import JSON_IMPORT_EXAMPLE
from hiking.models import Hike
from hiking.utils import DATA_HOME, DEFAULT_BOX_STYLE, SlimDateRange

# TODO: find a way to auto-detect this from rich
BOX_FORMATS = [
    "ASCII",
    "ASCII2",
    "ASCII_DOUBLE_HEAD",
    "SQUARE",
    "SQUARE_DOUBLE_HEAD",
    "MINIMAL",
    "MINIMAL_HEAVY_HEAD",
    "MINIMAL_DOUBLE_HEAD",
    "SIMPLE",
    "SIMPLE_HEAD",
    "SIMPLE_HEAVY",
    "HORIZONTALS",
    "ROUNDED",
    "HEAVY",
    "HEAVY_EDGE",
    "HEAVY_HEAD",
    "DOUBLE",
    "DOUBLE_EDGE",
    "MARKDOWN",
]


def get_valid_fields_for_args(exclude: Optional[List] = None):
    exclude = exclude or []
    return [
        field.info["name"]
        for field in Hike.FIELDS
        if field.info["data_view"] and field.info["name"] not in exclude
    ]


class DateRangeType:
    description = (
        "Only include hikes contained in provided daterange (default: all hikes)"
    )
    examples = (
        "Valid examples:\n"
        "1970-01-01\n"
        "1970-01-01/1970-02-01\n"
        "1970-01-01/ (all hikes from start date)\n"
        "/1970-01-01 (all hikes until end date)"
    )
    help = f"{description}\n{examples}"

    def __call__(self, raw: str, *args, **kwargs):
        start = end = raw
        splitted = raw.split("/")
        if len(splitted) == 2 and all(splitted):
            start, end = splitted
        elif start.endswith("/"):
            start = start.rstrip("/")
            end = None
        elif end.startswith("/"):
            end = end.lstrip("/")
            start = None

        try:
            start = (
                datetime.datetime.strptime(start, "%Y-%m-%d").date()
                if start
                else datetime.date.min
            )
            end = (
                datetime.datetime.strptime(end, "%Y-%m-%d").date()
                if end
                else datetime.date.max
            )
        except ValueError as e:
            raise argparse.ArgumentTypeError(f"{e.args[0]}\n{self.examples}")
        return SlimDateRange(start, end)


class WritableDirPathType:
    def __call__(self, raw: str, *args, **kwargs):
        directory = Path(raw)
        if (
            not directory.exists()
            or not directory.is_dir()
            or not os.access(directory, os.W_OK)
        ):
            raise argparse.ArgumentTypeError(
                f'Cannot write to directory: "{directory.absolute()}". '
                f"Make sure it exists and is writable."
            )
        return directory


class GPXFileType(argparse.FileType):
    def __call__(self, *args, **kwargs):
        file = super().__call__(*args, **kwargs)
        try:
            gpx_xml = file.read()
            gpxpy.parse(gpx_xml)
        except Exception as e:
            raise argparse.ArgumentTypeError(f"Cannot read *.gpx file: {str(e)}")
        return gpx_xml


class JsonFileType(argparse.FileType):
    def __call__(self, *args, **kwargs):
        file = super().__call__(*args, **kwargs)
        try:
            data = json.load(file)
        except Exception as e:
            raise argparse.ArgumentTypeError(f"Cannot read *.json file: {str(e)}")
        return data


def validate_order_key(value: str) -> Tuple[str, bool]:
    reverse = False
    if value.startswith("-"):
        reverse = True
        value = value.lstrip("-")
    if value not in get_valid_fields_for_args():
        raise argparse.ArgumentTypeError("Invalid order_key")
    return value, reverse


def validate_plot(value: str) -> Tuple[str, str]:
    try:
        x, y = tuple(value.split(","))
        for i in x, y:
            assert i in get_valid_fields_for_args(["name"])
        return x, y
    except (ValueError, AssertionError):
        raise argparse.ArgumentTypeError("plot")


def validate_table_style(value: str) -> Tuple[str, str]:
    try:
        box_style = getattr(box, value.upper())
    except AttributeError:
        raise argparse.ArgumentTypeError("Invalid table-style")
    return box_style


def set_default_subparser(
    parser: argparse.ArgumentParser, default_subcommand: str, raw_args: List[str]
):
    subparser_found = False
    for arg in raw_args:
        if arg in ["-h", "--help"]:  # pragma: no cover
            break
    else:
        for x in parser._subparsers._actions:
            if not isinstance(x, argparse._SubParsersAction):
                continue
            for sp_name in x._name_parser_map.keys():
                if sp_name in raw_args:
                    subparser_found = True
        if not subparser_found:
            raw_args.insert(0, default_subcommand)


def parse_arguments(raw_args: List[str]) -> argparse.Namespace:
    """
    Parse all arguments.
    """

    def format_list(data: List[str]):
        data.sort()
        col_1 = data[: int(round(len(data) / 2))]
        col_2 = data[int(round(len(data) / 2)) :]
        table_data = list(zip_longest(col_1, col_2, fillvalue=""))
        longest = 0
        for i in table_data:
            if len(i[0]) > longest:
                longest = len(i[0])
        result = [f"{i[0].ljust(longest)}{' ' * 5}{i[1]}" for i in table_data]
        return "\n".join(result)

    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawTextHelpFormatter, prog="hiking"
    )

    debug_arg_dict = {
        "help": "Show debug information (log queries)",
        "action": "store_true",
    }

    subparsers = parser.add_subparsers(dest="command")

    show = subparsers.add_parser(
        "show",
        help="Show hike(s) (default)",
        description="Show hike(s) (default)",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    show.add_argument(
        "ids",
        metavar="ID",
        help="Hike ID",
        nargs="*",
        type=int,
    )

    show.add_argument(
        "-d",
        "--daterange",
        help=DateRangeType.help,
        type=DateRangeType(),
        default=SlimDateRange(datetime.date.min, datetime.date.max),
    )

    show.add_argument(
        "-s",
        "--search",
        help="Search for text in name and body (case insensitive)",
        type=str,
    )

    show.add_argument(
        "-t",
        "--table-style",
        help=(
            "Table format style (default: simple)\n"
            f"Available options:\n{format_list(BOX_FORMATS)}"
        ),
        default=DEFAULT_BOX_STYLE,
        type=validate_table_style,
    )

    show.add_argument(
        "-o",
        "--order-key",
        help=(
            'Key to use for hike sorting. To reverse, prepend with "-".\n'
            "Available options:\n"
            f"{format_list(get_valid_fields_for_args())}"
        ),
        default=("date", False),
        type=validate_order_key,
    )

    show.add_argument(
        "--plot",
        help=(
            "Fields to plot in a graph.\n"
            "*experimental*\n"
            "Example:\n"
            '"date,distance"\n'
            "Available options:\n"
            f"{format_list(get_valid_fields_for_args(exclude=['name']))}"
        ),
        default=(),
        type=validate_plot,
    )

    show.add_argument("--debug", **debug_arg_dict)

    create = subparsers.add_parser(
        "create",
        help="Create a new record",
        description="Create a new record.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    create.add_argument(
        "--gpx",
        metavar="GPX_FILE",
        type=GPXFileType("r"),
        help="Import from *.gpx-file",
    )

    create.add_argument("--debug", **debug_arg_dict)

    edit = subparsers.add_parser(
        "edit",
        help="Edit a record",
        description="Edit a record.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    edit.add_argument(
        "id",
        metavar="ID",
        help="Hike ID",
        type=int,
    )

    edit.add_argument(
        "--gpx",
        metavar="GPX_FILE",
        type=GPXFileType("r"),
        help="Import from *.gpx-file",
    )

    edit.add_argument("--debug", **debug_arg_dict)

    delete = subparsers.add_parser(
        "delete",
        help="Delete records by ID",
        description="Delete records by ID.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    delete.add_argument(
        "-f",
        "--force",
        help="Do not ask before deletion",
        action="store_true",
    )

    delete.add_argument(
        "-q",
        "--quiet",
        help="Do not display hikes before deletion",
        action="store_true",
    )

    delete.add_argument(
        "-a",
        "--all",
        help="Delete all hikes",
        action="store_true",
    )

    delete.add_argument(
        "ids",
        metavar="ID",
        help="Hike ID",
        nargs="*",
        type=int,
    )

    delete.add_argument("--debug", **debug_arg_dict)

    _import = subparsers.add_parser(
        "import",
        help="Import records from JSON",
        description=f"Import records from JSON.\nFormat:\n{JSON_IMPORT_EXAMPLE}",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    _import.add_argument(
        "json_data",
        metavar="JSON_FILE",
        help="Path to JSON file",
        type=JsonFileType("r"),
    )

    _import.add_argument("--debug", **debug_arg_dict)

    export = subparsers.add_parser(
        "export",
        help="Export records as JSON and GPX",
        description="Export records as JSON and GPX.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    export.add_argument(
        "export_dir",
        metavar="EXPORT_DIR",
        help="Path to export directory",
        type=WritableDirPathType(),
    )

    export.add_argument(
        "ids",
        metavar="ID",
        help="Hike ID",
        nargs="*",
        type=int,
    )

    export.add_argument(
        "-d",
        "--daterange",
        help=DateRangeType.help,
        type=DateRangeType(),
        default=SlimDateRange(datetime.date.min, datetime.date.max),
    )

    export.add_argument(
        "-i",
        "--include-ids",
        help='Include IDs in export. Needed for "update", must be omitted for "create"',
        action="store_true",
    )

    export.add_argument("--debug", **debug_arg_dict)

    set_default_subparser(parser, "show", raw_args)

    args = parser.parse_args(raw_args)

    if args.command == "delete" and not args.ids and not args.all:
        raise parser.error("IDs or --all must be provided")
    elif args.command == "delete" and args.ids and args.all:
        raise parser.error("Ambiguous argument: IDs and --all provided")

    try:
        WritableDirPathType()(DATA_HOME.parent)
    except argparse.ArgumentTypeError:
        raise parser.error(f"Cannot write to user data director: {DATA_HOME.parent}")

    return args
