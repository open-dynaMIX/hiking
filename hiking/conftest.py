from pathlib import Path

import pytest
from pytest_factoryboy.fixture import register

from hiking import factories
from hiking.models import create_tables, engine, session

OWN_DIR = Path(__file__).resolve().parent
GPX_TEST_FILE = OWN_DIR / "tests" / "data" / "test.gpx"


register(factories.HikeFactory)


@pytest.fixture(scope="session", autouse=True)
def setup_db():
    create_tables()


@pytest.fixture(autouse=True)
def db():
    yield
    with engine.connect() as con:
        con.execute("DELETE FROM hikes;")
    session.commit()
    session.close()
    session.begin()


@pytest.fixture
def gpx_xml():
    with GPX_TEST_FILE.open("r") as f:
        xml = f.read()
    return xml
