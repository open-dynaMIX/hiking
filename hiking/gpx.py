from typing import TYPE_CHECKING

from hiking.plot import plot

if TYPE_CHECKING:  # pragma: no cover
    from models import Hike


def get_elevation_profile(hike: "Hike"):
    if not hike.gpx:
        return "No *.gpx-file available"

    x = []
    y = []
    for point in hike.gpx.get_points_data():
        x.append(point.distance_from_start / 1000)
        y.append(round(point.point.elevation))

    return plot(x, y, x_limit_min=0, xlabel="Distance (km)", ylabel="elevation (m)")
