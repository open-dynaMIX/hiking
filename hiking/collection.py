import datetime
from dataclasses import dataclass
from typing import List, Tuple, Union

from sqlalchemy import Interval, func
from sqlalchemy.orm import Query

from hiking.db_utils import session
from hiking.models import Hike, get_filtered_query
from hiking.utils import SlimDateRange, format_value


@dataclass
class HikeCollection:
    hikes: Query

    def get_hikes_attr_list(
        self, attr: str
    ) -> List[Union[float, int, datetime.date, datetime.timedelta]]:
        return [getattr(hike, attr) for hike in self.hikes.all()]

    def sum(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        is_interval = isinstance(getattr(Hike, attr).expression.type, Interval)
        if Hike.FIELD_PROPS[attr]["calculated_value"] or is_interval:
            attr_list = self.get_hikes_attr_list(attr)
            if is_interval:
                return sum(attr_list, datetime.timedelta())

            return sum(attr_list)  # pragma: no cover

        result = session.query(func.sum(getattr(Hike, attr))).first()[0]
        return result

    def avg(self, attr: str = "distance") -> Union[float, datetime.timedelta]:
        if attr == "duration":
            return (
                sum([h.duration for h in self.hikes.all()], datetime.timedelta())
                / self.hikes.count()
            )
        return sum(getattr(h, attr) for h in self.hikes.all()) / self.hikes.count()

    def max(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        if attr == "speed":
            return sorted(self.hikes.all(), key=lambda x: x.speed, reverse=True)[
                0
            ].speed

        result = getattr(
            self.hikes.order_by(getattr(Hike, attr).desc()).limit(1).first(), attr
        )
        return result

    def min(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        if attr == "speed":
            return sorted(self.hikes.all(), key=lambda x: x.speed)[0].speed

        result = getattr(
            self.hikes.order_by(getattr(Hike, attr).asc()).limit(1).first(), attr
        )
        return result

    def get_hikes_stats(self, order_params: Tuple[str, bool]) -> List[List[str]]:
        if order_params[0] == "speed":
            hike_list = sorted(
                self.hikes.all(), key=lambda x: x.speed, reverse=order_params[1]
            )
            return [hike.get_stats() for hike in hike_list]

        order = getattr(Hike, order_params[0])
        if order_params[1]:
            order = order.desc()
        query = self.hikes.order_by(order)
        return [hike.get_stats() for hike in query]

    def calc_and_format_value(self, calc: str, attr: str) -> str:
        result = getattr(self, calc)(attr)
        return format_value(result, attr)

    def get_totals(self) -> List[str]:
        def get_summary_cell(attr: str, supported_calculations: List[str]):
            if supported_calculations:
                cell = {}
                for calc, pretty_calc in [
                    ("sum", "Σ "),
                    ("avg", "⌀ "),
                    ("max", "↑ "),
                    ("min", "↓ "),
                ]:
                    cell[pretty_calc] = "-"
                    if calc in supported_calculations:
                        cell[pretty_calc] = self.calc_and_format_value(calc, attr)
                return cell

        d = [
            "",
            "STATS",
            str(self.hikes.count()),
            *[
                get_summary_cell(attr, config["supported_calculations"])
                for attr, config in Hike.FIELD_PROPS.items()
                if config["supported_calculations"]
            ],
        ]
        return d

    def get_collection_stats(
        self, order_params: Tuple[str, bool], add_totals: bool = True
    ) -> Tuple[List, List]:
        stats = self.get_hikes_stats(order_params)

        result = [
            *stats,
        ]

        footer = None
        if add_totals and len(stats) > 1:
            # Only add totals if more than one hike is present
            footer = self.get_totals()

        return result, footer

    def __str__(self) -> str:
        return (
            f"<HikeCollection - "
            f"containing {self.hikes.count()} hikes - "
            f"total {self.sum()}km>"
        )

    def __repr__(self) -> str:
        return self.__str__()


def get_collection(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = get_filtered_query(ids, daterange)
    collection = HikeCollection(hikes=query)
    return collection
