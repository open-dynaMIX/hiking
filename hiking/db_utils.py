from typing import List

from hiking.collection import HikeCollection
from hiking.models import Hike, session
from hiking.utils import SlimDateRange


def get_filtered_query(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = (
        session.query(Hike)
        .filter(Hike.date >= daterange.lower)
        .filter(Hike.date <= daterange.upper)
    )
    if ids:
        query = query.filter(Hike.id.in_(ids))
    return query


def get_collection(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = get_filtered_query(ids, daterange)
    collection = HikeCollection(hikes=query, session=session)
    return collection
