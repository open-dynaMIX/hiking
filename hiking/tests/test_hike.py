import datetime

import pytest
from gpxpy.gpx import GPX

from hiking.gpx import get_elevation_profile
from hiking.models import Hike, session


def test_hike(known_hike, snapshot):
    assert known_hike.speed == 2.8617511520737327
    assert known_hike.get_pretty_value("speed") == "2.86"

    assert known_hike.get_stats() == snapshot
    assert known_hike.get_detail_stats() == snapshot

    query = session.query(Hike)
    assert query.count() == 1
    assert session.query(Hike).first().name == known_hike.name
    known_hike.name = "new name"
    known_hike.save()
    assert session.query(Hike).first().name == "new name"


def test_hike_load_gpx(hike, gpx_xml):
    hike.load_gpx(gpx_xml)
    assert isinstance(hike.gpx, GPX)
    assert hike.name == hike.gpx.name
    assert hike.date == hike.gpx.time.date()
    assert hike.distance == round(hike.gpx.length_3d() / 1000, 2)
    assert hike.elevation_gain == round(hike.gpx.get_uphill_downhill().uphill)
    assert hike.elevation_loss == round(hike.gpx.get_uphill_downhill().downhill)
    assert hike.gpx_xml == gpx_xml
    # duration differs, because our test gpx file has no timestamps
    hike.save()


@pytest.mark.parametrize("add_gpx", [True, False])
def test_hike_gpx_elevation_profile(add_gpx, hike, gpx_xml, snapshot):
    if add_gpx:
        hike.gpx_xml = gpx_xml
        hike.save()

    ele_profile = get_elevation_profile(hike)
    assert ele_profile == snapshot


def test_hike_save_delete():
    hike = Hike(
        name="Foo",
        date=datetime.date(1999, 9, 9),
        distance=23.5,
        elevation_gain=123,
        elevation_loss=666,
        duration=datetime.timedelta(minutes=90),
    )
    q = session.query(Hike)
    assert q.count() == 0
    hike.save()
    assert q.count() == 1
    assert hike == q.first()
    hike.name = "Bar"
    hike.save()
    assert q.count() == 1
    assert q.first().name == "Bar"
    hike.delete()
    assert q.count() == 0
