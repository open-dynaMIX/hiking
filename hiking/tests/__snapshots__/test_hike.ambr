# serializer version: 1
# name: test_hike
  list([
    '1',
    '2010-05-20',
    'Paul Diaz',
    '17.4',
    '3268',
    '2652',
    '03:50',
    '4.53',
  ])
# ---
# name: test_hike.1
  dict({
    'Date': datetime.date(2010, 5, 20),
    'ID': 1,
    'Name': 'Paul Diaz',
    'km/h': 4.53,
    '⏱ ': '03:50',
    '➡ km': 17.43,
    '⬈ m': 3268,
    '⬊ m': 2652,
  })
# ---
# name: test_hike_gpx_elevation_profile[False]
  'No *.gpx-file available'
# ---
# name: test_hike_gpx_elevation_profile[True]
  '''
  (elevation (m)) ^
         557 |
         557 | ⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         556 | ⡇⠀⠀⠀⡤⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         555 | ⡇⠀⠀⡸⠀⠈⢆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         554 | ⡇⠀⢠⠃⠀⠀⠈⢆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         553 | ⡇⠀⡸⠀⠀⠀⠀⠘⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         552 | ⡇⠀⡇⠀⠀⠀⠀⠀⢇⠀⡜⢆⡸⠉⠉⢣⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         552 | ⡇⢠⠃⠀⠀⠀⠀⠀⠘⠚⠀⠈⠃⠀⠀⠀⠓⠢⡀⠀⠀⡰⠲⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         551 | ⡇⢸⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⠒⠚⠀⠀⢣⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⠖⠒⠒⠒⠒⠀⠀⠀⠀
         550 | ⡇⢸⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠦⠤⠤⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⡎⠀⠀⠀⠀⠀⠀⠀⠀⠀
         549 | ⡇⡎⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠸⡀⠀⠀⢀⣀⠀⠀⠀⢰⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀
         548 | ⡇⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢇⠀⠀⢸⠀⢣⠀⠀⡜⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         547 | ⡇⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠉⠉⠉⠀⠀⠉⠉⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         546 | ⣇⠇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         546 | ⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         545 | ⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         544 | ⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         543 | ⡏⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         542 | ⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         541 | ⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
         541 | ⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
  -----------|-|---------|---------|---------|---------|--------> (Distance (km))
             | 0.00      0.09      0.18      0.27      0.36     
  '''
# ---
